### PR TITLE
QCQMC part 3: Add Hamiltonian

### PR DESCRIPTION
**Pull request by fdmalone**
Adds hamiltonian module for specifying a second quantized hamiltonian used for optimizing a trial wavefunction. 

Depends on #342 and #343.

cc @mpharrigan

### CODE REVIEW

**Comment by fdmalone** [comment]
@mpharrigan / @wjhuggins PTAL, the diff should be a lot more digestible now! The only change over the private repo is to use attrs instead of dataclasses.

**Comment by fdmalone** [comment]
Also this doesn't depend on quaff yet, so I didn't touch the CI.

**Comment by fdmalone** [comment]
@dstrain115 / @mpharrigan PTAL

File: recirq/qcqmc/hamiltonian.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
+from pathlib import Path
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:1]: This file is missing the license.  I noticed this is part 3.  Did we miss licenses in parts 1 and 2?

File: recirq/qcqmc/hamiltonian.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+from typing import Tuple, Union
+
+import attrs
+import fqe
+import h5py
+import numpy as np
+import openfermion as of
+from fqe.hamiltonians.restricted_hamiltonian import RestrictedHamiltonian
+from fqe.openfermion_utils import integrals_to_fqe_restricted
+from openfermionpyscf._run_pyscf import compute_integrals  # type: ignore
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:11]: Why are we importing from a private module?  This is generally not advisable.

File: recirq/qcqmc/hamiltonian.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+from pyscf import ao2mo, fci, gto, scf
+
+from recirq.qcqmc.config import OUTDIRS
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:14]: Use import statements for packages and modules only, not for individual types, classes, or functions.

https://github.com/qh-lab/pyle/pull/new/u/dstrain/minor_uwaves_speed_up

File: recirq/qcqmc/hamiltonian.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+from recirq.qcqmc.data import Data, Params, get_integrals_path
+
+
+@attrs.frozen
+class LoadFromFileHamiltonianParams(Params):
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:19]: This name is awkward.  Load is a verb, so this seems like an action, but the class holds parameters?
I would rename to HamiltonianFileParameters

File: recirq/qcqmc/hamiltonian.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    """Class for storing the parameters that specify loading integrals from a file."""
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:20]: Missing docstring about the vairables, such as n_orb, etc.

File: recirq/qcqmc/hamiltonian.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+
+    @property
+    def path_string(self, prefix: str = "") -> str:
+        return self.path_prefix + OUTDIRS.DEFAULT_HAMILTONIAN_DIRECTORY + self.name
+
+
+# xyz coordinates of an atom
+PositionT = Tuple[float, float, float]
+# [atom name, xyz]
+GeomT = Tuple[str, PositionT]
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:40]: nit: I would put constants at the top above the class.

File: recirq/qcqmc/hamiltonian.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    def chk_path(self) -> Path:
+        """Path string to any checkpoint file is saved."""
+        parent = self.base_path.parent
+        if not parent.is_dir():
+            parent.mkdir(exist_ok=True, parents=True)
+        return self.base_path.with_suffix(".chk")
+
+    @property
+    def pyscf_molecule(self) -> gto.Mole:
+        """Underlying pyscf mol instance."""
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:101]: mol -> molecule.  Avoid abbreviations, especially when they are ambiguous.

File: recirq/qcqmc/hamiltonian.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+        return molecule
+
+
+HamiltonianParams = Union[LoadFromFileHamiltonianParams, PyscfHamiltonianParams]
+
+
+@attrs.frozen
+class HamiltonianData(Data):
+    """Data class that contains information about a Hamiltonian."""
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:123]: add docstring for the arguments.

File: recirq/qcqmc/hamiltonian_test.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
+import numpy as np
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian_test.py:1]: Add license here too.

File: recirq/qcqmc/hamiltonian.py (hunk 3 above)
Comment (mpharrigan) [reply recirq/qcqmc/hamiltonian.py:14]: did you mean to include that link?

That's a google style guide thing, right? we don't follow that in ReCirq and just use pep8

File: recirq/qcqmc/hamiltonian.py (hunk 7 above)
Comment (mpharrigan) [reply recirq/qcqmc/hamiltonian.py:101]: I think it's pyscf doing the abbreviating here, but they seem to call it `Mole`

Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:101]: Is that the animal or the sauce? I'll update to Mole

File: recirq/qcqmc/hamiltonian.py (hunk 3 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:14]: on a separate note would it be worthwhile (or possible) to add incremental formatting / lint checks for recirq?

File: recirq/qcqmc/hamiltonian.py (hunk 8 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:123]: Done.

File: recirq/qcqmc/hamiltonian.py (hunk 6 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:40]: Done.

File: recirq/qcqmc/hamiltonian.py (hunk 5 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:20]: Done.

File: recirq/qcqmc/hamiltonian.py (hunk 7 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:101]: Done.

File: recirq/qcqmc/hamiltonian_test.py (hunk 9 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian_test.py:1]: Done.

Comment (fdmalone) [reply recirq/qcqmc/hamiltonian_test.py:1]: Done.

File: recirq/qcqmc/hamiltonian.py (hunk 4 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:19]: Done

File: recirq/qcqmc/hamiltonian.py (hunk 1 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:15]: 1 was ok, 2 was the external (vendored) dependency so I just added a license to the directory (but not to the individual files). 4 was missing the license which I added here.

File: recirq/qcqmc/hamiltonian.py (hunk 3 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py]: Oh it is linting with flake nvm

File: recirq/qcqmc/hamiltonian.py (hunk 2 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py]: I copied the openfermion function into hamiltonian.py, WDYT?

**Comment by fdmalone** [comment]
@dstrain115 PTAL

File: recirq/qcqmc/hamiltonian.py [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+        return (
+            self.path_prefix + config.OUTDIRS.DEFAULT_HAMILTONIAN_DIRECTORY + self.name
+        )
+
+
+@attrs.frozen
+class PyscfHamiltonianParams(data.Params):
+    """Class for describing the parameters to get a Hamiltonian from Pyscf.
+
+    name: A name for the Hamiltonian
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:112]: Indentation is a little messed up here.  Should begin with "Args:" and the rest should be indented 4 spaces.

File: recirq/qcqmc/hamiltonian.py [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+
+        molecule.build()
+
+        assert int(molecule.nao_nr()) == self.n_orb
+        assert molecule.nelectron == self.n_elec
+
+        return molecule
+
+
+HamiltonianParams = Union[HamiltonianFileParams, PyscfHamiltonianParams]
```
Comment (dstrain115) [inline recirq/qcqmc/hamiltonian.py:185]: Let's put this at the top too.

**Review by dstrain115 — APPROVED** [review]

File: recirq/qcqmc/hamiltonian.py (hunk 10 above)
Comment (fdmalone) [reply recirq/qcqmc/hamiltonian.py:112]: Whoops

**Comment by fdmalone** [comment]
@mpharrigan / @dstrain115 could you merge I don't have permission

**Pull request merged**